attentive-android-sdk/src/main/java/com/attentive/androidsdk/AttentiveApi.java

```diff
@@ -0,0 +1,194 @@
+            @Override
+            public void onFailure(@NonNull Call call, @NonNull IOException e) {
+                callback.onFailure("Getting geo-adjusted domain failed: " + e.getMessage());
+            }
+
+            @Override
+            public void onResponse(@NonNull Call call, @NonNull Response response) throws IOException {
+                // Check explicitly for 200 (instead of response.isSuccessful()) because the response only has the tag in the body when its code is 200
+                if (response.code() != 200) {
+                    callback.onFailure(String.format("Getting geo-adjusted domain returned invalid code: '%d', message: '%s'", response.code(), response.message()));
+                    return;
+                }
+
+                ResponseBody body = response.body();
+
+                if (body == null) {
+                    callback.onFailure("Getting geo-adjusted domain returned no body");
+                    return;
+                }
+
+                String fullTag = response.body().string();
+                String geoAdjustedDomain = parseAttentiveDomainFromTag(fullTag);
+
+                if (geoAdjustedDomain == null) {
+                    callback.onFailure("Could not parse the domain from the full tag");
+                    return;
+                }
+
+                internalSendUserIdentifiersCollectedEventAsync(geoAdjustedDomain, userIdentifiers, callback);
+            }
+        });
+    }
+
+    private void internalSendUserIdentifiersCollectedEventAsync(String geoAdjustedDomain, UserIdentifiers userIdentifiers, AttentiveApiCallback callback) {
+        String externalVendorIdsJson = null;
+        try {
+            List<ExternalVendorId> externalVendorIds = buildExternalVendorIds(userIdentifiers);
+            externalVendorIdsJson = objectMapper.writeValueAsString(externalVendorIds);
+        } catch (JsonProcessingException e) {
+            callback.onFailure(String.format("Could not serialize the UserIdentifiers. Message: '%s'", e.getMessage()));
+            return;
+        }
+
+        String metadataJson = null;
+        try {
+            Metadata metadata = buildMetadata(userIdentifiers);
+            metadataJson = objectMapper.writeValueAsString(metadata);
+        } catch (JsonProcessingException e) {
+            callback.onFailure(String.format("Could not serialize metadata. Message: '%s'", e.getMessage()));
+            return;
+        }
+
+        HttpUrl.Builder urlBuilder = getHttpUrlEventsEndpointBuilder()
+            .addQueryParameter("v", "mobile-app")
+            .addQueryParameter("c", geoAdjustedDomain)
+            .addQueryParameter("t", "idn")
+            .addQueryParameter("evs", externalVendorIdsJson)
+            .addQueryParameter("m", metadataJson);
+
+        if (userIdentifiers.getVisitorId() != null) {
+            urlBuilder.addQueryParameter("u", userIdentifiers.getVisitorId());
+        }
```

Comment:
if the visitor id is null should we generate one?

Reply:
Let me think about this. This should technically never happen :/, because AttentiveConfig always sets the visitorId on the UserIdentifier. We just can't mark visitorId as NonNull because the host app also uses the UserIdentifier to pass the other non-VisitorId ids to AttentiveConfig. So this is a code smell that I'm not sure yet how to fix. Maybe separating into two objects: one for the host app to pass identifiers to our SDK, and other that we can use internally in the SDK (which will include the VisitorId).

Reply:
hmmm yeah maybe we can have a `UserIdentity` object that stores the visitor ID and the UserIdentifiers? and the visitor ID is not null? 

no I think we can address in another PR!

Reply:
yeah, I like that idea. sg